utils/frame/rpc/system/src/lib.rs

```diff
@@ -41,207 +38,219 @@ use sp_block_builder::BlockBuilder;
+	Pool: TransactionPool + 'static,
+	Fetcher: light::Fetcher<Block> + 'static,
 	Block: traits::Block,
 	AccountId: Clone + std::fmt::Display + Codec + Send + 'static,
 	Index: Clone + std::fmt::Display + Codec + Send + traits::AtLeast32Bit + 'static,
 {
-	fn nonce(&self, account: AccountId) -> FutureResult<Index> {
-		todo!();
-		/*let best_hash = self.client.info().best_hash;
+	async fn nonce(&self, account: AccountId) -> Result<Index, JsonRpseeCallError> {
+		let best_hash = self.client.info().best_hash;
 		let best_id = BlockId::hash(best_hash);
-		let future_best_header = future_header(&*self.remote_blockchain, &*self.fetcher, best_id);
-		let fetcher = self.fetcher.clone();
+		let best_header = future_header(&*self.remote_blockchain, &*self.fetcher, best_id)
+			.await
+    		.map_err(|blockchain_err| JsonRpseeCallError::Failed(Box::new(blockchain_err)))?
+			.ok_or_else(|| ClientError::UnknownBlock(format!("{}", best_hash)))
+    		.map_err(|client_err| JsonRpseeCallError::Failed(Box::new(client_err)))?;
 		let call_data = account.encode();
-		let future_best_header = future_best_header
-			.and_then(move |maybe_best_header| ready(
-				maybe_best_header.ok_or_else(|| { ClientError::UnknownBlock(format!("{}", best_hash)) })
-			));
-		let future_nonce = future_best_header.and_then(move |best_header|
-			fetcher.remote_call(RemoteCallRequest {
+		let nonce = self.fetcher.remote_call(RemoteCallRequest {
 				block: best_hash,
 				header: best_header,
 				method: "AccountNonceApi_account_nonce".into(),
 				call_data,
 				retry_count: None,
 			})
-		).compat();
-		let future_nonce = future_nonce.and_then(|nonce| Decode::decode(&mut &nonce[..])
-			.map_err(|e| ClientError::CallResultDecode("Cannot decode account nonce", e)));
-		let future_nonce = future_nonce.map_err(|e| RpcError {
-			code: ErrorCode::ServerError(Error::RuntimeError.into()),
-			message: "Unable to query nonce.".into(),
-			data: Some(format!("{:?}", e).into()),
-		});
-
-		let pool = self.pool.clone();
-		let future_nonce = future_nonce.map(move |nonce| adjust_nonce(&*pool, account, nonce));
-
-		Box::new(future_nonce)*/
+			.await
+			.map_err(|blockchain_err| JsonRpseeCallError::Failed(Box::new(blockchain_err)))?;
+
+		let nonce: Index = Decode::decode(&mut &nonce[..])
+			.map_err(|codec_err| JsonRpseeCallError::Failed(Box::new(codec_err)))?;
+
+		Ok(adjust_nonce(&*self.pool, account, nonce))
 	}
 
-	fn dry_run(&self, _extrinsic: Bytes, _at: Option<<Block as traits::Block>::Hash>) -> FutureResult<Bytes> {
-		todo!();
-		// Box::new(result(Err(RpcError {
-		//     code: ErrorCode::MethodNotFound,
-		//     message: "Unable to dry run extrinsic.".into(),
-		//     data: None,
-		// })))
+	async fn dry_run(&self, _extrinsic: Bytes, _at: Option<<Block as traits::Block>::Hash>) -> Result<Bytes, JsonRpseeCallError> {
+		Err(JsonRpseeCallError::Custom {
+			code: -32601, // TODO: (dp) We have this in jsonrpsee too somewhere. This is jsonrpsee::ErrorCode::MethodNotFound
```

Comment:
yes, it's lower down in the stack but we can add this variant to the call error I guess.